lib/ruby_indexer/lib/ruby_indexer/configuration.rb

```diff
@@ -60,21 +80,46 @@ def indexables
       # NOTE: indexing the patterns (both included and excluded) needs to happen before indexing gems, otherwise we risk
       # having duplicates if BUNDLE_PATH is set to a folder inside the project structure
 
+      flags = File::FNM_PATHNAME | File::FNM_EXTGLOB
+
+      # In order to speed up indexing, only traverse into top-level directories that are not entirely excluded.
+      # For example, if "tmp/**/*" is excluded, we don't need to traverse into "tmp" at all. However, if
+      # "vendor/bundle/**/*" is excluded, we will traverse all of "vendor" and `reject!` out all "vendor/bundle" entries
+      # later.
+      included_paths = Dir.glob(File.join(@workspace_path, "*/"), flags)
+        .reject { |dir| File.fnmatch?(merged_excluded_file_pattern, dir, flags) }
+        .map do |included_path|
+          relative_path = included_path
+            .delete_prefix(@workspace_path)
+            .tap { |path| path.delete_prefix!("/") }
+            .tap { |path| path.delete_suffix!("/") }
+
+          [included_path, relative_path]
+        end
```

Comment:
Two things about this section:

1. We can avoid looping twice by using `filter_map`
2. We should invoke `merged_excluded_file_pattern` only once. It's currently invoking it once per directory, which is more expensive than necessary

Also, why do we need to delete the `/` suffix twice?

```suggestion
      excluded_pattern = merged_excluded_file_pattern
      included_paths = Dir.glob(File.join(@workspace_path, "*/"), flags)
        .filter_map do |included_path|
          next if File.fnmatch?(excluded_pattern, dir, flags)
          
          relative_path = included_path
            .delete_prefix(@workspace_path)
            .tap { |path| path.delete_prefix!("/") }
            .tap { |path| path.delete_suffix!("/") }

          [included_path, relative_path]
        end
```

Reply:
Yeah, I just removed that line. We want prefixing `/` gone, but we want to keep the suffix ones.
https://github.com/Shopify/ruby-lsp/pull/2082/commits/14495ade396c2772f29e8a97bee1eb9f1aca9558

Reply:
The only reason we didn't have any tests failing is because we weren't including a path whose top-level directory started with an excluded directory name.